### PR TITLE
Asynchronous ArcHostContext serializations

### PR DESCRIPTION
**Pull request by ianull**
This PR works together with [cl/324891905](http://cl/324891905) to improve Arcs startup time 400+ ms

### CODE REVIEW

File: java/arcs/core/host/AbstractArcHost.kt [hunk 1]
```diff
@@ -99,8 +105,31 @@ abstract class AbstractArcHost(
...
+
+        /**
+         * Asynchronous [ArcHostContext] serializations in observed order in a hot [Flow].
+         * The [channel] will be cancelled when the [updateArcHostContextScope] is cancelled.
+         *
+         * TODO:
+         * make the channel per-Arc instead of per-Host for better serialization
+         * performance under multiple running and to-be-run Arcs.
+         */
+        channel.consumeAsFlow().onEach { it() }.launchIn(updateArcHostContextScope)
```
Comment (jblebrun) [inline java/arcs/core/host/AbstractArcHost.kt:123]: Why do we need a flow here?

For a simple case like this, just using the channel directly seems fine:
```
val contextSerializationJob = scope.launch {
   channel.forEach { it() }
}
```

File: java/arcs/core/host/AbstractArcHost.kt [hunk 2]
```diff
@@ -99,8 +105,31 @@ abstract class AbstractArcHost(
     // TODO: add lifecycle API for ArcHosts shutting down to cancel running coroutines
     private val scope = CoroutineScope(coroutineContext)
 
+    /** Supports asynchronous [ArcHostContext] serializations. */
+    private val channel: Channel<suspend () -> Unit> = Channel(Channel.UNLIMITED)
+    private val updateArcHostContextScope = CoroutineScope(updateArcHostContextCoroutineContext)
```
Comment (jblebrun) [inline java/arcs/core/host/AbstractArcHost.kt:110]: Is it possible to instead pass in a scope instead of creating one here?

File: java/arcs/core/host/AbstractArcHost.kt [hunk 3]
```diff
@@ -312,10 +344,36 @@ abstract class AbstractArcHost(
...
+            )
+            channel.send {
+                try {
+                    contextCopy?.let {
+                        /** TODO: reuse [ArcHostContextParticle] instances if possible. */
+                        createArcHostContextParticle(it).run {
+                            writeArcHostContext(it.arcId, it)
+                        }
+                        // Remove the reference in the flow/channel cache.
+                        contextCopy = null
```
Comment (jblebrun) [inline java/arcs/core/host/AbstractArcHost.kt:364]: Why is this needed? Where is this being cached?

File: java/arcs/core/host/AbstractArcHost.kt (hunk 1 above)
Comment (ianull) [reply java/arcs/core/host/AbstractArcHost.kt:123]: What is `channel.forEach`?

The channel receiver side needs to do something like
```
scope.launch { while (true) channel.receive()() }
```
This is equivalent to using a hot flow to collect/receive element from the channel.
```
channel.consumeAsFlow().onEach { it() }.launchIn(scope)
``` 

I'm ok with the two styles and the latter looks a bit better to me.

File: java/arcs/core/host/AbstractArcHost.kt (hunk 2 above)
Comment (ianull) [reply java/arcs/core/host/AbstractArcHost.kt:110]: It makes sense to pass in a scope if no shutdown nor lifecycle binding can be observed in the `AbstractArcHost`.

The `shutdown()` in the `AbstractArcHost` already takes care of scope cancellation for the `scope` created from the passed-in `coroutineContext `, so I just did the same for `updateArcHostContextScope`.

Any problem you foresee creating a scope in the `AbstractArcHost` and cancelling it in `AbstractArcHost.shutdown`?

File: java/arcs/core/host/AbstractArcHost.kt (hunk 3 above)
Comment (ianull) [reply java/arcs/core/host/AbstractArcHost.kt:364]: Please see:
https://github.com/PolymerLabs/arcs/pull/5352
https://github.com/PolymerLabs/arcs/pull/5413

The reference is held by `result in FlowKt_ChannelsKt...`

File: java/arcs/core/host/AbstractArcHost.kt (hunk 1 above)
Comment (jblebrun) [reply java/arcs/core/host/AbstractArcHost.kt:123]: a ReceiveChannel is iterable, you should be able to do:
```
scope.launch {
  for(task in channel) {
    task()
  }
}
```

To me the intent here is much clearer than with the flow.

File: java/arcs/core/host/AbstractArcHost.kt (hunk 3 above)
Comment (jblebrun) [reply java/arcs/core/host/AbstractArcHost.kt:364]: This seems like another argument against flows. Will this issue occur if we're just using channels?

File: java/arcs/core/host/AbstractArcHost.kt (hunk 1 above)
Comment (ianull) [reply java/arcs/core/host/AbstractArcHost.kt:123]: hmm ok, i'll change it if it's clearer to read.

Comment (jblebrun) [reply java/arcs/core/host/AbstractArcHost.kt:123]: I am willing to keep debating and be swayed in the other direction. 

I get the value of flows for the cold stream usage, but using them to wrap channels for simple scenarios like this feels less valuable to me.

For this *particular* use case, introducing flows feels like additional complexity and overhead for no perceived benefit... but it's definitely possible that I'm missing or forgetting something.

File: java/arcs/core/host/AbstractArcHost.kt [hunk 4]
```diff
@@ -60,6 +65,7 @@ typealias ParticleRegistration = Pair<ParticleIdentifier, ParticleConstructor>
 @ExperimentalCoroutinesApi
 abstract class AbstractArcHost(
     protected val coroutineContext: CoroutineContext = Dispatchers.Default,
+    updateArcHostContextCoroutineContext: CoroutineContext = Dispatchers.Default,
```
Comment (jblebrun) [inline java/arcs/core/host/AbstractArcHost.kt:68]: Is the main goal of adding this parameter so that we could change to a different dispatcher if needed?

File: java/arcs/core/host/AbstractArcHost.kt (hunk 2 above)
Comment (jblebrun) [reply java/arcs/core/host/AbstractArcHost.kt:110]: I'm OK leaving it as is for now, I think we have general work to do on improving our scope creation/passing strategy project-wide. 

The `launchIn` of the flow will return a job that you can cancel on shutdown, so there's no need to save a scope.

File: java/arcs/core/host/AbstractArcHost.kt (hunk 1 above)
Comment (ianull) [reply java/arcs/core/host/AbstractArcHost.kt:123]: For simple use case, channel or flow (wrap a channel) doesn't make big difference to me.
The flow just provides much flexibility like combining flows, flowOn to switch coroutine context, onCompletion to take care the tear-down jobs upon cancellation, etc.

For now, only send and receive-then-execute the task, so changing to use channel directly sounds reasonable to me.
In the future, if other modules need to listen to i.e. completion of serializations, a broadcasted channel as a flow is required for easier flow combination.

File: java/arcs/core/host/AbstractArcHost.kt (hunk 3 above)
Comment (ianull) [reply java/arcs/core/host/AbstractArcHost.kt:364]: It only occurs in flows. iirc at flow's `emitAll` function.

File: java/arcs/core/host/AbstractArcHost.kt (hunk 4 above)
Comment (ianull) [reply java/arcs/core/host/AbstractArcHost.kt:68]: Yes, like the cl that is waiting for this PR: [cl/324891905](http://cl/324891905)

File: java/arcs/core/host/AbstractArcHost.kt (hunk 2 above)
Comment (jblebrun) [reply java/arcs/core/host/AbstractArcHost.kt:110]: I think you should leave as-is for now, since it's similar to other patterns we have. 

But, I think this might be in the spirit of Kotlin structured concurrency:
```
AbstractArcHost(
  coroutineScope: CoroutineScope
  arcSerializationDispatcher: CoroutineDispatcher
  ...
) {
  ...
  init {
  ...
   channel.consumeAsFlow().onEach { it() }.launchIn(coroutineScope + arcSerializationDispatcher)
  }
```

This way, there's still just a single scope for structure, but we can use the right dispatcher for the channel/flow/etc.

File: java/arcs/core/host/AbstractArcHost.kt (hunk 1 above)
Comment (ianull) [reply java/arcs/core/host/AbstractArcHost.kt]: Refactored to use Channel directly for the time being.

File: java/arcs/core/host/AbstractArcHost.kt (hunk 2 above)
Comment (ianull) [reply java/arcs/core/host/AbstractArcHost.kt]: LGTM to central mgmt on scopes, pass in one scope and optionally one or multiple dispatchers.

File: java/arcs/core/host/AbstractArcHost.kt [hunk 5]
```diff
@@ -105,29 +119,29 @@ abstract class AbstractArcHost(
...
-         * make the channel per-Arc instead of per-Host for better serialization
-         * performance under multiple running and to-be-run Arcs.
-         */
-        channel.consumeAsFlow().onEach { it() }.launchIn(updateArcHostContextScope)
     }
 
     /** Wait until all observed context serializations got flushed. */
     private suspend fun drainSerializations() {
-        if (!channel.isClosedForSend) {
+        if (!contextSerializationChannel.isClosedForSend) {
```
Comment (jblebrun) [inline java/arcs/core/host/AbstractArcHost.kt:142]: Is this check needed? The channel never gets closed right?

File: java/arcs/core/host/AbstractArcHost.kt [hunk 6]
```diff
@@ -211,7 +225,7 @@ abstract class AbstractArcHost(
         runningMutex.withLock { runningArcs.clear() }
         clearContextCache()
         pausedArcs.clear()
-        updateArcHostContextScope.cancel()
+        contextSerializationTasks.cancel()
```
Comment (jblebrun) [inline java/arcs/core/host/AbstractArcHost.kt:228]: would we want to attempt to let the channel drain here? Something like...

```
contextSerializationChannel.close()
contextSerializationTasks.join()
```

File: java/arcs/core/host/AbstractArcHost.kt [hunk 7]
```diff
@@ -346,22 +360,18 @@ abstract class AbstractArcHost(
      */
     protected open suspend fun writeContextToStorage(arcId: String, context: ArcHostContext) {
         /** Serialize the [context] to storage in observed order asynchronously. */
-        if (!channel.isClosedForSend) {
-            var contextCopy: ArcHostContext? = ArcHostContext(
+        if (!contextSerializationChannel.isClosedForSend) {
```
Comment (jblebrun) [inline java/arcs/core/host/AbstractArcHost.kt:363]: If we close the channel in ArcHost shutdown, wouldn't it be an error to arrive here with the channel closed? so we should just let it crash?

File: java/arcs/core/host/AbstractArcHost.kt (hunk 5 above)
Comment (jblebrun) [reply java/arcs/core/host/AbstractArcHost.kt:142]: If we close in shutdown it could be, but I have a similar question to below:
If we close the channel in ArcHost shutdown, wouldn't it be an error to arrive here with the channel closed? so we should just let it crash?

Comment (ianull) [reply java/arcs/core/host/AbstractArcHost.kt:142]: a failed channel: a closed channel due to exceptions.
so used to add a isClosedFor{Send|Receive} check.

File: java/arcs/core/host/AbstractArcHost.kt (hunk 6 above)
Comment (ianull) [reply java/arcs/core/host/AbstractArcHost.kt:228]: I thought about it, but don't think it's a good idea here at tear-down flow to block for a long time.
A context serialization takes 400-500 ms and if there are multiple of them, we might cause ANR if the tear-down is run at app lifecycle like onDestroy.

Comment (ianull) [reply java/arcs/core/host/AbstractArcHost.kt:228]: corrections: currently the pause() function at the shutdown() would drain all serialization tasks before it actually shuts everything down. So the join() is not necessary and we might hit ANR if the serialization tasks really take a lot of time.

File: java/arcs/core/host/AbstractArcHost.kt (hunk 7 above)
Comment (ianull) [reply java/arcs/core/host/AbstractArcHost.kt:363]: crash is too bruteforced for a background serialization task.
I think a better way is just fallback.
```
val task = suspend { ... }

if (!contextSerializationChannel.isClosedForSend) channel.send(task)
else task()
```

Comment (ianull) [reply java/arcs/core/host/AbstractArcHost.kt:369]: Refactored to support fallback!

File: java/arcs/core/host/AbstractArcHost.kt [hunk 8]
```diff
@@ -68,7 +71,24 @@ abstract class AbstractArcHost(
...
         schedulerProvider: SchedulerProvider,
         vararg initialParticles: ParticleRegistration
-    ) : this(Dispatchers.Default, schedulerProvider, null, *initialParticles)
+    ) : this(Dispatchers.Default, Dispatchers.Default, schedulerProvider, null, *initialParticles)
+
+    /**
+     * Backward-compatible for some existing [AbstractArcHost] implementations to dispatch
+     * [contextSerializationTasks] onto the [coroutineContext] instead of separate context.
+     */
+    constructor(
```
Comment (jblebrun) [inline java/arcs/core/host/AbstractArcHost.kt:80]: Can we mark this as `@Deprecated`, including a `replaceWith` param?

File: java/arcs/core/host/AbstractArcHost.kt [hunk 9]
```diff
@@ -99,10 +119,33 @@ abstract class AbstractArcHost(
...
+    /**
+     * Supports asynchronous [ArcHostContext] serializations in observed order.
+     *
+     * TODO:
+     * make the channel per-Arc instead of per-Host for better serialization
+     * performance under multiple running and to-be-run Arcs.
+     */
+    private val contextSerializationChannel: Channel<suspend () -> Unit> =
+        Channel(Channel.UNLIMITED)
+    private val contextSerializationTasks =
```
Comment (jblebrun) [inline java/arcs/core/host/AbstractArcHost.kt:131]: `contextSerializationJob` may be a better name, since the returned type is `Job`.

File: java/arcs/core/host/AbstractArcHost.kt (hunk 7 above)
Comment (jblebrun) [reply java/arcs/core/host/AbstractArcHost.kt:369]: This is not a background task, this is the scheduling of a background task. 

The only way the channel will be closed is if we have closed it, which currently occurs in `shutdown`.

If `shutdown` has been called on the ArcHost, it means that the component owning the ArcHost has called `shutdown` on the Arc, but then explicitly called `startArc` or `stopArc` after that. This is incorrect usage, and should throw an exception.

File: java/arcs/core/host/AbstractArcHost.kt (hunk 5 above)
Comment (jblebrun) [reply java/arcs/core/host/AbstractArcHost.kt:142]: This looks to me like it masks potential developer errors.

We have full control over when the channel is closed (currently, only due to `cancel` in the `shutdown` method).

We have full control over when we attempt to send to it (occurs due to an outside call to `startArc`, `stopArc`, or `pause`).

My assertion here is that if we call `shutdown`, there should not be subsequent interaction with the ArcHost that results in sending to the channel.

File: java/arcs/core/host/AbstractArcHost.kt (hunk 7 above)
Comment (ianull) [reply java/arcs/core/host/AbstractArcHost.kt:369]: I actually see things differently: a startArc/stopArc is happening during ArcHost shutdown.
The shutdown has gone over the line of channel.close and racing startArc/stopArc just arrives at the checkpoint of isClosedForSend. I would like to choose to fallback instead of throwing an exception right away as the startArc/stopArc is happening during shutdown should be made not effective rather than throwing an exception.

As for the sequential case: a startArc/stopArc after a completed shutdown.
The assertion/exception/check should be put in the entry of startArc/stopArc rather than in the middle way of them.

File: java/arcs/core/host/AbstractArcHost.kt (hunk 8 above)
Comment (ianull) [reply java/arcs/core/host/AbstractArcHost.kt:80]: sure! lemme add it

File: java/arcs/core/host/AbstractArcHost.kt (hunk 9 above)
Comment (ianull) [reply java/arcs/core/host/AbstractArcHost.kt:131]: ok! will rename it, thanks

File: java/arcs/core/host/AbstractArcHost.kt (hunk 7 above)
Comment (jblebrun) [reply java/arcs/core/host/AbstractArcHost.kt:369]: I think we should revisit this outside of this PR: Contractually, one should not call action methods on a component after closing it/shutting it down. Shutdown is a terminal operation. 

Calling `stopArc` or `startArc` after calling `shutdown` is a programmer error, so we should identify it and eliminate it.

Comment (ianull) [reply java/arcs/core/host/AbstractArcHost.kt:393]: Yep, we can revisit it later.
I agree with shutdown IS a terminal op.

However a startArc can be started prior to a shutdown and they run at different threads racing in the ArcHost.
This is the situation I mentioned.

Currently on G3, only chronicle on->off will call the shutdown, so I think it's safe now.

Comment (jblebrun) [reply java/arcs/core/host/AbstractArcHost.kt:393]: Ok I see your point.
